File: apps/server/src/jobs/steps/discover.ts
import { appendFileSync } from 'node:fs'

interface DockerPsEntry {
  ID: string
  Image: string
  Names: string
}

// Exported for unit testing — pure parsing helpers with no side effects.

export function extractImageName(imageTag: string): string {
  // Strip only the tag portion (after the last `/`, find the first `:`)
  // Preserves port numbers in registry hostnames, e.g. localhost:5001/image:tag → localhost:5001/image
  const lastSlash = imageTag.lastIndexOf('/')
  const afterLastSlash = imageTag.slice(lastSlash + 1)
  const tagStart = afterLastSlash.indexOf(':')
  if (tagStart < 0)
    return imageTag
  return imageTag.slice(0, lastSlash + 1 + tagStart)
}

export function findMatchingContainerId(
  psOutput: string,
  imageName: string,
): { id: string, name: string } | null {
  for (const line of psOutput.split('\n')) {
    if (!line.trim())
      continue
    let entry: DockerPsEntry
    try {
      entry = JSON.parse(line) as DockerPsEntry
    }
    catch {
      continue
    }
    // Match containers whose image is exactly `imageName` (bare) or `imageName:tag`
    if (entry.Image === imageName || entry.Image.startsWith(`${imageName}:`)) {
      return { id: entry.ID, name: entry.Names.replace(/^\//, '') }
    }
  }
  return null
}

export function extractComposeInfo(
  inspectOutput: string,
  containerName: string,
): { composePath: string, service: string } {
  let labels: Record<string, string> | null
  try {
    labels = JSON.parse(inspectOutput.trim()) as Record<string, string> | null
  }
  catch {
    throw new Error(`Container ${containerName} returned invalid JSON from docker inspect`)
  }

  if (!labels)
    throw new Error(`Container ${containerName} has no Docker labels — not started via docker compose`)

  // config_files may list multiple comma-separated paths when using -f overrides; take the first
  const composePath = (labels['com.docker.compose.project.config_files'] ?? '').split(',')[0]?.trim()
  const service = labels['com.docker.compose.service']

  if (!composePath)
    throw new Error(`Container ${containerName} is missing 'config_files' label — not started via docker compose`)

  if (!service)
    throw new Error(`Container ${containerName} is missing 'service' label — not started via docker compose`)

  return { composePath, service }
}

export async function discover(imageTag: string, app: string, logPath: string): Promise<{ composePath: string, service: string }> {
  const log = (line: string) => appendFileSync(logPath, `[${new Date().toISOString()}] ${line}\n`)
  const imageName = extractImageName(imageTag)

  log(`[discover] Searching for containers using image: ${imageName}`)

  const psProc = Bun.spawn(['docker', 'ps', '--format', '{{json .}}'], {
    stdout: 'pipe',
    stderr: 'pipe',
  })

  const [psExit, psStdout, psStderr] = await Promise.all([
    psProc.exited,
    new Response(psProc.stdout).text(),
    new Response(psProc.stderr).text(),
  ])

  if (psExit !== 0)
    throw new Error(`docker ps failed (exit ${psExit}): ${psStderr}`)

  const match = findMatchingContainerId(psStdout, imageName)

  if (!match) {
    log(`[discover] No running containers found matching image: ${imageName}`)
    log(`[discover] Hint: Ensure the service was started at least once before using rollhook`)
    log(`[discover] Hint: Run 'docker ps' on your server to verify the container is running`)
    log(`[discover] Hint: The image registry prefix must match exactly what Docker shows for the running container`)
    log(`[discover] Hint: To verify: docker ps --filter ancestor=${imageName} (check image names match)`)
    throw new Error(`No running container found matching image: ${imageName}`)
  }

  log(`[discover] Found container: ${match.name} (ID: ${match.id.slice(0, 12)})`)

  const inspectProc = Bun.spawn(['docker', 'inspect', match.id, '--format', '{{json .Config.Labels}}'], {
    stdout: 'pipe',
    stderr: 'pipe',
  })

  const [inspectExit, inspectStdout, inspectStderr] = await Promise.all([
    inspectProc.exited,
    new Response(inspectProc.stdout).text(),
    new Response(inspectProc.stderr).text(),
  ])

  if (inspectExit !== 0)
    throw new Error(`docker inspect failed (exit ${inspectExit}): ${inspectStderr}`)

  const { composePath, service } = extractComposeInfo(inspectStdout, match.name)

  log(`[discover] Compose file: ${composePath}`)
  log(`[discover] Service: ${service}`)
  log(`[discover] Discovery complete`)

  return { composePath, service }
}

File: internal/jobs/steps/discover.go
package steps

import (
	"context"
	"errors"
	"fmt"
	"strings"

	"github.com/docker/docker/api/types/container"
	"github.com/docker/docker/client"

	dockerpkg "github.com/jkrumm/rollhook/internal/docker"
)

// ErrServiceNotFound is returned by Discover when no running container matches the image.
// Callers can use errors.Is to distinguish this from infrastructure errors.
var ErrServiceNotFound = errors.New("service not found")

// DiscoveryResult holds Compose metadata extracted from a running container.
type DiscoveryResult struct {
	ComposePath string
	Service     string
	Project     string
	Labels      map[string]string // all labels from the discovered container
}

// Discover finds the running container whose image matches imageTag and
// extracts its Docker Compose metadata (compose file path, service, project).
func Discover(ctx context.Context, cli *client.Client, imageTag string) (*DiscoveryResult, error) {
	imageName := ExtractImageName(imageTag)

	containers, err := dockerpkg.ListRunningContainers(ctx, cli)
	if err != nil {
		return nil, fmt.Errorf("discover: %w", err)
	}

	match := FindMatchingContainer(containers, imageName)
	if match == nil {
		return nil, fmt.Errorf("%w: no running container found matching image: %s", ErrServiceNotFound, imageName)
	}

	detail, err := dockerpkg.InspectContainer(ctx, cli, match.ID)
	if err != nil {
		return nil, fmt.Errorf("discover: %w", err)
	}

	return ExtractComposeInfo(detail.Config.Labels, containerName(match))
}

// ExtractImageName strips the tag from an image reference, preserving
// registry host:port prefixes.
//
//	"localhost:5000/app:v1" → "localhost:5000/app"
//	"nginx:latest" → "nginx"
//	"nginx" → "nginx"
func ExtractImageName(imageTag string) string {
	lastSlash := strings.LastIndex(imageTag, "/")
	afterLastSlash := imageTag[lastSlash+1:]
	tagStart := strings.Index(afterLastSlash, ":")
	if tagStart < 0 {
		return imageTag
	}
	return imageTag[:lastSlash+1+tagStart]
}

// FindMatchingContainer returns the first container whose Image field matches
// imageName. Matching is tried in order: exact name match, then base-name
// fallback (last path segment) to handle registry-prefixed images when
// searching by bare name (e.g. "myapp" matches "registry.example.com/myapp").
func FindMatchingContainer(containers []container.Summary, imageName string) *container.Summary {
	searchBase := baseName(imageName)
	for i := range containers {
		c := &containers[i]
		cName := ExtractImageName(c.Image)
		if cName == imageName || baseName(cName) == searchBase {
			return c
		}
	}
	return nil
}

// baseName returns the last path segment of an image name (after the final slash).
func baseName(image string) string {
	if idx := strings.LastIndex(image, "/"); idx >= 0 {
		return image[idx+1:]
	}
	return image
}

// ExtractComposeInfo reads Docker Compose labels from a container's label map
// and returns the compose file path, service name, and project name.
func ExtractComposeInfo(labels map[string]string, name string) (*DiscoveryResult, error) {
	if labels == nil {
		return nil, fmt.Errorf("container %s has no Docker labels — not started via docker compose", name)
	}

	configFiles := labels["com.docker.compose.project.config_files"]
	composePath := strings.TrimSpace(strings.SplitN(configFiles, ",", 2)[0])
	if composePath == "" {
		return nil, fmt.Errorf("container %s is missing 'config_files' label — not started via docker compose", name)
	}

	service := labels["com.docker.compose.service"]
	if service == "" {
		return nil, fmt.Errorf("container %s is missing 'service' label — not started via docker compose", name)
	}

	project := labels["com.docker.compose.project"]
	if project == "" {
		return nil, fmt.Errorf("container %s is missing 'project' label — not started via docker compose", name)
	}

	return &DiscoveryResult{ComposePath: composePath, Service: service, Project: project, Labels: labels}, nil
}

func containerName(c *container.Summary) string {
	if len(c.Names) == 0 {
		return c.ID[:min(12, len(c.ID))]
	}
	return strings.TrimPrefix(c.Names[0], "/")
}
